api/src/main/java/redux/api/Store.java

```diff
@@ -0,0 +1,102 @@
+package redux.api;
+
+/**
+ * Coordinates actions and [Reducers][Reducer]. Store has the following responsibilities:
+ * <ul>
+ * 	<li>Holds application state</li>
+ * 	<li>Allows access to state via {@link #getState()}</li>
+ * 	<li>Allows state to be updated via {@link Dispatcher#dispatch(Object)}</li>
+ * 	<li>Registers listeners via {@link #subscribe(Subscriber)}</li>
+ * 	<li>Handles unregistering of listeners via the {@link Subscriber} returned by {@link #subscribe(Subscriber)}</li>
+ * </ul>
+ *
+ * @see <a href="http://redux.js.org/docs/basics/Store.html">http://redux.js.org/docs/basics/Store.html</a>
+ */
+public interface Store<S> extends Dispatcher {
+
+	/**
+	 * Returns the current state tree of your application. It is equal to the last value returned by the store’s
+	 * reducer.
+	 *
+	 * @return the current state
+	 * @see <a href="http://redux.js.org/docs/api/Store.html#getState">http://redux.js.org/docs/api/Store.html#getState</a>
+	 */
+	S getState();
+
+	/**
+	 * Adds a change listener. It will be called any time an action is dispatched, and some part of the state tree may
+	 * potentially have changed. You may then call {@link #getState()} to read the current state tree inside the callback.
+	 *
+	 * @return A subscription
+	 * @param subscriber The subscriber
+	 * @see <a href="http://redux.js.org/docs/api/Store.html#subscribe">http://redux.js.org/docs/api/Store.html#subscribe</a>
+	 */
+	Subscription subscribe(Subscriber subscriber);
+
+	/**
+	 * Replaces the reducer currently used by the store to calculate the state.
+	 *
+	 * @param reducer The reducer
+	 * @see <a href="http://redux.js.org/docs/api/Store.html#replaceReducer">http://redux.js.org/docs/api/Store.html#replaceReducer</a>
+	 */
+	void replaceReducer(Reducer<S> reducer);
+
+	/**
+	 * An interface that creates a store.
+	 *
+	 * @see <a href="http://redux.js.org/docs/Glossary.html#store-creator">http://redux.js.org/docs/Glossary.html#store-creator</a>
+	 */
+	interface Creator<S> {
+
+		/**
+		 * Creates a store.
+		 *
+		 * @param reducer Reducer for the store state
+		 * @param initialState Initial state for the store
+		 * @return The store
+		 */
+		Store<S> create(Reducer<S> reducer, S initialState);
```

Comment:
@beyondeye, @glung and I just chatted about this gist: https://gist.github.com/pardom/21124eca191b092ed1a886e2452f00a6

It demonstrates the type enforcement I was referring to, while still allowing for an enhanced State/Store.